vald/evm/evm.go

```diff
@@ -431,7 +433,27 @@ func decodeEventContractCallWithToken(log *geth.Log) (types.EventContractCallWit
 	}, nil
 }
 
-func (mgr Mgr) getTxReceiptIfFinalized(chain nexus.ChainName, txID common.Hash, confHeight uint64) (*geth.Receipt, error) {
+func (mgr Mgr) isTxReceiptFinalized(chain nexus.ChainName, txReceipt *geth.Receipt, confHeight uint64) (bool, error) {
+	client, ok := mgr.rpcs[strings.ToLower(chain.String())]
+	if !ok {
+		return false, fmt.Errorf("rpc client not found for chain %s", chain.String())
+	}
+
+	if mgr.latestFinalizedBlockCache.Get(chain).Cmp(txReceipt.BlockNumber) >= 0 {
+		return true, nil
+	}
+
+	isFinalized, err := client.IsFinalized(context.Background(), confHeight, txReceipt)
+	if err != nil || !isFinalized {
+		return isFinalized, err
+	}
+
+	mgr.latestFinalizedBlockCache.Set(chain, txReceipt.BlockNumber)
```

Comment:
We should cache the actual finalized block (which is determined as part of IsTxFinalized, for Eth2 like chains, it'll be the block returned by the finalized tag, for confirmation based ones, it'll be the latest block - conf height, etc.). So I'm thinking that we can pass the cache to that method, or have the caching be per client. Otherwise this optimization isn't going to be useful for txs in different blocks since contract calls are usually triggered in order of block number.

Reply:
While it does make sense, no the cache should never be passed to the RPC client. I'll change `IsFinalized` to `GetLatestFinalizedBlockNumber` though.

Reply:
Yep. For some chains, we might not get the latest finalized block number vs checking if a given block is finalized, but we can change the naming if and when needed for a chain.